### PR TITLE
Add azd compatibility

### PR DESCRIPTION
**Pull request by wangmingliang-ms**

### CODE REVIEW

File: README.md
Comment (jongio) [inline README.md]: Add http links.

File: README.md
Comment (jongio) [inline README.md]: What do you mean by 'pass them'?  They can load the .env file into their app context using a version of the java dotenv lib right?

File: src/main/resources/application-azure.properties
Comment (weikanglim) [inline src/main/resources/application-azure.properties:3]: Do you need to do this? This was only used by the java ToDO because we do have private configuration stored in Azure KeyVault.. I don't see any code changes in the Java application that has changed to reference new secrets

File: infra/database/mysql.bicep
Comment (jongio) [inline infra/database/mysql.bicep]: Can you please also do a PR to get this into azure-dev bicep core?

File: infra/main.bicep
Comment (jongio) [inline infra/main.bicep]: dependsOn shouldn't be required.

File: infra/main.bicep
Comment (jongio) [inline infra/main.bicep]: You shouldn't need any dependsOn if the other resource is referenced by properties.  Can you scrub all bicep and remove dependsOn?

File: src/main/resources/application-azure.properties
Comment (wangmingliang-ms) [reply src/main/resources/application-azure.properties:3]: the `MYSQL_PASS` is stored in Azure KeyVault.

https://github.com/Azure-Samples/spring-petclinic-java-mysql/blob/825d7b4a6c7ab41e37d38ea27dfdcdb06854615b/src/main/resources/application-mysql.properties#L5

https://github.com/wangmingliang-ms/spring-petclinic-java-mysql/blob/7fcd9d8d10285b6d388721fa5c94087a330a4279/infra/database/mysql.bicep#L97-L103
```bicep
resource mysqlAdminPasswordSecret 'Microsoft.KeyVault/vaults/secrets@2022-07-01' = {
  parent: keyVault
  name: mysqlAdminPassKey
  properties: {
    value: mysqlAdminPassword
  }
}
```

https://microsoft.github.io/spring-cloud-azure/current/reference/html/index.html#secret-management

File: infra/main.bicep
Comment (wangmingliang-ms) [reply infra/main.bicep]: sure, I will remove them.

File: infra/app/appserviceplan.bicep
Comment (jongio) [inline infra/app/appserviceplan.bicep]: Please use what we have in our templates under /infra/core and put those files in your project under infra core and use /infra/app for app specific files.

File: README.md
Comment (wangmingliang-ms) [reply README.md]: added

File: infra/database/mysql.bicep
Comment (wangmingliang-ms) [reply infra/database/mysql.bicep]: yes, PR is here https://github.com/Azure/azure-dev/pull/1257

**Comment by jongio** [comment]
Please let me know when ready for re-review

File: README.md
Comment (weikanglim) [inline README.md]: `azd` currently requires Java 17 SDK and above. I think it's generally a good idea for templates to be up-to-date, so I think it makes sense to bump this up here.

There's a question of whether that's too limiting from an `azd` perspective, but given that we're targeting newer audience, I think the strong requirement is fine for now. But let me know if you think otherwise.

File: README.md
Comment (weikanglim) [inline README.md]: Removing the extra bullet points:

```suggestion
> manually via the Azure portal or with the `azd down` command.
```

File: README.md
Comment (weikanglim) [inline README.md]:
```suggestion
> `--template` parameter the first time you call it to get the code locally. Subsequent `azd up` calls do not require the template
> parameter. If you do provide the parameter, all your local source code will be overwritten if you agree to overwrite when prompted.
```

File: README.md
Comment (weikanglim) [inline README.md]: @wangmingliang-ms Do you think it'd be better if we move down to before the Security section? That way it reads more like a readme -- which assumes the user doesn't know much. Then we can explain the differences.

Comment (weikanglim) [inline README.md]: Do you mean to link to the files in the commit? This currently loads the diff between `~9ecdc11..9ecdc11`.

Does this link make more sense? https://github.com/spring-projects/spring-petclinic/tree/9ecdc1111e3da388a750ace41a125287d9620534

File: README.md
Comment (weikanglim) [inline README.md]: What do you think about being more explicit here?

```suggestion
* Added `./infra` folder and `./azure.yaml` `azd` project configuration: application "infrastructure as code" files to create and configure Azure resources.
```

File: README.md
Comment (weikanglim) [inline README.md]:
```suggestion
* Modified `./pom.xml` and `./build.gradle`: import required libraries for `Azure Key Vault`/`Azure Monitor`(`Azure Application Insights`) integration.
```

File: README.md
Comment (weikanglim) [inline README.md]: For the image to be rendered in the bullet point in strict markdown, need an extra blank line:

```suggestion
    * `pom.xml`

      ![diff of pom.xml](readme.assests/pom.xml.diff.png)
    * `build.gradle`

      ![diff of build.gradle](readme.assests/build.gradle.diff.png)
```

File: README.md
Comment (weikanglim) [inline README.md]:
```suggestion
* `./src/main/java/org/springframework/samples/petclinic/PetClinicApplication.java`: enable monitoring and logging with `Azure Application Insights`

  ![diff of PetClinicApplication.java](readme.assests/PetClinicApplication.java.diff.png)
```

File: src/main/resources/application-azure.properties
Comment (weikanglim) [reply src/main/resources/application-azure.properties:3]: Make sense. I don't have permissions to resolve this comment, so will leave it up to you.

File: src/main/java/org/springframework/samples/petclinic/PetClinicApplication.java
Comment (weikanglim) [inline src/main/java/org/springframework/samples/petclinic/PetClinicApplication.java:35]: There's a recent change in App Service Java17 container that breaks programmatic attach. For the time being, you'll need to add this change in the `infra` folder for `app-service`: https://github.com/Azure/azure-dev/pull/1217/files

**Comment by weikanglim** [comment]
Looks good to me, pending remaining comments.

File: src/main/java/org/springframework/samples/petclinic/PetClinicApplication.java
Comment (weikanglim) [reply src/main/java/org/springframework/samples/petclinic/PetClinicApplication.java:35]: Might be solved if you just use the unreleased version of [api-appservice-java.bicep](https://github.com/Azure/azure-dev/pull/1217/files#diff-b4dd64ba60ffa7851a99f4256a0cba8dc3ce5af6f113d0e39070a9fbc88f7769)

Comment (wangmingliang-ms) [reply src/main/java/org/springframework/samples/petclinic/PetClinicApplication.java:35]: I ran into this error some days ago, but didn't know how to fix it at that time, thanks a lot!
```
2022-12-13T02:41:51.333845501Z Caused by: io.opentelemetry.contrib.attach.RuntimeAttachException: A problem has occurred during the runtime attachment of the Java agent.
2022-12-13T02:41:51.333849101Z 	at io.opentelemetry.contrib.attach.CoreRuntimeAttach.handleByteBuddyException(CoreRuntimeAttach.java:105)
2022-12-13T02:41:51.333852501Z 	at io.opentelemetry.contrib.attach.CoreRuntimeAttach.attachJavaagentToCurrentJVM(CoreRuntimeAttach.java:54)
2022-12-13T02:41:51.333856002Z 	at com.microsoft.applicationinsights.attach.ApplicationInsights.attach(ApplicationInsights.java:61)
2022-12-13T02:41:51.333859402Z 	at org.springframework.samples.petclinic.PetClinicApplication.main(PetClinicApplication.java:35)
2022-12-13T02:41:51.333862902Z 	... 8 more
2022-12-13T02:41:51.333866202Z Caused by: java.lang.IllegalStateException: Could not self-attach to current VM using external process
2022-12-13T02:41:51.333869802Z 	at net.bytebuddy.agent.ByteBuddyAgent.installExternal(ByteBuddyAgent.java:700)
2022-12-13T02:41:51.333873202Z 	at net.bytebuddy.agent.ByteBuddyAgent.install(ByteBuddyAgent.java:631)
2022-12-13T02:41:51.333876602Z 	at net.bytebuddy.agent.ByteBuddyAgent.attach(ByteBuddyAgent.java:298)
2022-12-13T02:41:51.333880002Z 	at net.bytebuddy.agent.ByteBuddyAgent.attach(ByteBuddyAgent.java:263)
2022-12-13T02:41:51.333883402Z 	at net.bytebuddy.agent.ByteBuddyAgent.attach(ByteBuddyAgent.java:245)
2022-12-13T02:41:51.333886803Z 	at io.opentelemetry.contrib.attach.CoreRuntimeAttach.attachJavaagentToCurrentJVM(CoreRuntimeAttach.java:52)
```

File: infra/app/appserviceplan.bicep
Comment (wangmingliang-ms) [reply infra/app/appserviceplan.bicep]: restructured the files by following the same pattern as azd-dev.

File: README.md
Comment (wangmingliang-ms) [reply README.md]: I don't have preference for this, I will fix as you suggested. thanks.

Comment (wangmingliang-ms) [reply README.md]: you are right, fixed, thanks.

File: README.md
Comment (wangmingliang-ms) [reply README.md]: I'd probably think in terms of the minimum version **recommended** by Azure App Service/Azure Functions/Azure Spring Apps. 
But as a user, I can understand if it requires at least Java17

File: README.md
Comment (wangmingliang-ms) [reply README.md]: there is such lib in java, but not applicable for azd project `.env` (which is usually not be packaged):
1. We have no way to get the path of the `.env` file **at runtime**, because we have no way to reference `project.baseDir` at runtime, while the .env file only exists in the project folder and it won't be packaged into the final jar. 
2. This `.env` does not exist in Azure environment because it won't be packaged into the final jar, so it seems redundant to introduce such a lib (it will be packaged into the final jar).

**Comment by wangmingliang-ms** [comment]
@weikanglim  @jongio the `azd provision` keeps `Compiling Bicep template...` at the last commit [306eb10](https://github.com/Azure-Samples/spring-petclinic-java-mysql/pull/1/commits/306eb1031dfd4af5e6b5bada0bfb26aeac44ed47) but OK at [6d306d3](https://github.com/Azure-Samples/spring-petclinic-java-mysql/pull/1/commits/6d306d313c9e7970268d38bb7c2c1a0047c67509)
do you know what is wrong? I just renamed some params or added some params in the last commit

image

File: infra/core/database/mysql/mysql-server.bicep
Comment (jongio) [inline infra/core/database/mysql/mysql-server.bicep]: Should we default this?  We do in the sqlserver.bicep file.

File: infra/core/database/mysql/mysql-server.bicep
Comment (jongio) [inline infra/core/database/mysql/mysql-server.bicep]: Can we spell out password here?

File: infra/core/database/mysql/mysql-server.bicep
Comment (jongio) [inline infra/core/database/mysql/mysql-server.bicep]: Change to `autoGrow`, not 'enableAutoGrow`

File: infra/main.bicep
Comment (jongio) [inline infra/main.bicep]: I believe we always use Sql, not SQL

File: README.md
Comment (jongio) [reply README.md]: I think we should help the user more here.  Like here's a script to load the vars, or something like that.  But, I would like @weikanglim  to chime in to see if he can think of a better way.

File: infra/core/database/mysql/mysql-server.bicep
Comment (wangmingliang-ms) [reply infra/core/database/mysql/mysql-server.bicep]: it's not the real password, but the key used to fetch the password from Key Vault.

File: infra/core/database/mysql/mysql-server.bicep
Comment (wangmingliang-ms) [reply infra/core/database/mysql/mysql-server.bicep]: fixed.

File: README.md
Comment (weikanglim) [reply README.md]: 1. If using `VSCode`, we have the `dotenv` task that will load the env file.
2. If looking at pure shell scripting, there are some ways to accomplish this. You can use `azd env get-values` to return the current selected `.env` file values, and use some lightweight scripts to load it into the shell.

## Bash
```bash
export $(azd env get-values | xargs -0)
```

## PowerShell:
```pwsh
azd env get-values | % { "`$env:$_".Replace('\"', '`"') } | Invoke-Expression
```

Comment (wangmingliang-ms) [reply README.md]: I added predefined tasks/configurations for VSCode users just same as other templates, and add more details about how to setup the environment variables in IntelliJ IDEA.

Comment (wangmingliang-ms) [reply README.md]: @jongio  @weikanglim  I added predefined tasks/configurations for VSCode users leveraging the `dotenv` task just same as other templates and also added more details about how to setup environment variables in IntelliJ IDEA.

**Comment by wangmingliang-ms** [comment]
@jongio @weikanglim  Hello, can I merge this PR now?

**Review by weikanglim — APPROVED** [review]
LGTM

**Comment by jongio** [comment]
Where are we at with adding mysql to /azd/bicep/core?

**Comment by wangmingliang-ms** [comment]
> Where are we at with adding mysql to /azd/bicep/core?

the PR is [here](https://github.com/Azure/azure-dev/pull/1257), all the comments were resolved. awaiting review from you.

**Pull request merged**
